compiler/stable_mir/src/ty.rs

```diff
@@ -565,3 +565,24 @@ pub enum ImplPolarity {
     Negative,
     Reservation,
 }
+
+#[derive(Clone, Debug)]
+pub struct Instance {
+    pub def: InstanceDef,
+    pub args: GenericArgs,
+}
+
+#[derive(Clone, Debug)]
+pub enum InstanceDef {
+    Item(DefId),
+    Intrinsic(DefId),
+    VTableShim(DefId),
+    ReifyShim(DefId),
+    FnPtrShim(DefId, Ty),
+    Virtual(DefId, usize),
+    ClosureOnceShim { call_once: DefId, track_caller: bool },
+    ThreadLocalShim(DefId),
+    DropGlue(DefId, Option<Ty>),
+    CloneShim(DefId, Ty),
+    FnPtrAddrShim(DefId, Ty),
+}
```

Comment:
I don't think this should expose all the kinds of shims rustc has. This list changes occasionally. Instead I think you should have `Item`, `Intrinsic`, `Virtual` and a generic `Shim` variant which only exposes just enough to get the Stable MIR for the specific instantiation of said shim.